### ISSUE TITLE
Comparing two recursive arrays leads to segfault

### ISSUE THREAD

**Issue by asterite** [crystal-lang/crystal#11837]
Consider this code:

```crystal
alias RecursiveArray = Array(RecursiveArray)

a = [] of RecursiveArray
a << a
puts a
puts a == a
```

It will give stack overflow.

I actually thought we implemented this in a way that it detects cycles, but it seems we never did that 😅 

Do you think we should do it? Here I used recursive arrays, but it could also happen if you have an array with an object that in turns holds a reference to the same array.

**Comment by asterite** [comment]
Of course doing this all the time slows down things, but we could include some optimizations. For example if an array is composed exclusively of primitive types there's no need to check for recursion. And my guess it that custom arrays are rarely compared in code, and when that happens, performance doesn't matter much.

**Comment by Fryguy** [comment]
As a developer I wouldn't expect the compiler to tell me about infinite recursion or infinite loops, so I'm tempted to say maybe we just shouldn't handle this.  Do you know what Rust or other compilers do in these cases?  If it's possible to detect though and it can be done in a way as to not cause that much overhead I'm not sure why we wouldn't do it.

At least you get a nice error currently, which from a dev perspective seems clear enough to fix.

```
[[...]]
Stack overflow (e.g., infinite or very deep recursion)
[0x109258fb7] *Exception::CallStack::print_backtrace:Nil +39 in /Users/jfrey/dev/./recurse
[0x109258f4f] ~procProc(Int32, Pointer(LibC::SiginfoT), Pointer(Void), Nil)@/usr/local/Cellar/crystal/1.3.0/src/signal.cr:127 +543 in /Users/jfrey/dev/./recurse
[0x7fff20932d7d] _sigtramp +29 in /usr/lib/system/libsystem_platform.dylib
[0x10925a401] *Array(RecursiveArray)@Array(T)#==<Array(RecursiveArray)>:Bool +65 in /Users/jfrey/dev/./recurse
[0x10925a405] *Array(RecursiveArray)@Array(T)#==<Array(RecursiveArray)>:Bool +69 in /Users/jfrey/dev/./recurse (140256452192810 times)
[0x109223aee] __crystal_main +2974 in /Users/jfrey/dev/./recurse
[0x109228b06] main +70 in /Users/jfrey/dev/./recurse
```

Aside that (140256452192810 times) is cracking me up 🤣

**Comment by asterite** [comment]
> As a developer I wouldn't expect the compiler to tell me about infinite recursion or infinite loops, so I'm tempted to say maybe we just shouldn't handle this.

This seems a bit contradictory. Could you explain it a bit further?

The idea is that right now the code I posted fails at runtime with stack overflow. I propose we change that so that it returns `true` (in this case,) making it work exactly like Ruby.

**Comment by Fryguy** [comment]
> > As a developer I wouldn't expect the compiler to tell me about infinite recursion or infinite loops, so I'm tempted to say maybe we just shouldn't handle this.
> 
> This seems a bit contradictory. Could you explain it a bit further?

I think I misunderstood your proposal and thought that you wanted the compiler to solve this during the compilation phase.  I don't think it's possible to solve this at compile time, as it would have to be a runtime check. So my statement was that if it were not possible to handle this at compile time, I think that's ok, because it's not something I would expect anyway.

> The idea is that right now the code I posted fails at runtime with stack overflow. I propose we change that so that it returns `true` (in this case,) making it work exactly like Ruby.

I'm continually amazed that after 15 years of Ruby, I learn new things.  I never knew Ruby handled recursive Objects during `==`.  I knew it did it during `.inspect`, so I shouldn't be surprised.  So yeah, if Ruby can do it, we absolutely also should do it.

(I coincidentally just read your dev.to articles on this subject...I really enjoyed them!)

**Comment by HertzDevil** [comment]
This is Ruby's implementation for `Array#==(Array)`: https://github.com/ruby/ruby/blob/c53bdb8ff64ad707ebbb0c05823c1cb0376433dc/array.c#L4964-L4977

which I think might be implemented as:

```crystal
class Reference
  # :nodoc:
  module ExecRecursivePaired
    alias Registry = Hash({UInt64, UInt64, Symbol}, Bool)

    {% if flag?(:preview_mt) %}
      @@exec_recursive = Crystal::ThreadLocalValue(Registry).new
    {% else %}
      @@exec_recursive = Registry.new
    {% end %}

    def self.hash
      {% if flag?(:preview_mt) %}
        @@exec_recursive.get { Registry.new }
      {% else %}
        @@exec_recursive
      {% end %}
    end
  end

  private def exec_recursive(method, arg)
    hash = ExecRecursivePaired.hash
    key = {object_id, arg.object_id, method}
    if hash[key]?
      false
    else
      hash[key] = true
      yield arg
      hash.delete(key)
      true
    end
  end
end

class Array(T)
  def ==(other : Array) : Bool
    exec_recursive(:equals?, other) do |arg|
      return false unless equals?(arg) { |x, y| x == y }
    end
    true
  end
end

alias Foo = Array(Foo)

a = [] of Foo
a << a
p a == a # => true

b = [] of Foo
b << b
p a == b # => true

c = [] of Foo
d = [] of Foo
d << c
c << d
p c == d # => true
```

[Here](https://github.com/ruby/ruby/blob/c53bdb8ff64ad707ebbb0c05823c1cb0376433dc/array.c#L4919) it seems a pair of arrays may be assumed to be equal if they are compared more than once. I don't know why.

**Comment by asterite** [comment]
> [Here](https://github.com/ruby/ruby/blob/c53bdb8ff64ad707ebbb0c05823c1cb0376433dc/array.c#L4919) it seems a pair of arrays may be assumed to be equal if they are compared more than once. I don't know why.

I've been thinking about that and I'm also not sure why. I think the reason is that it returns `true` when we encounter this pair of objects when we already started comparing them before. We don't know what will happen with the rest of the elements, but maybe it's okay to return `true` there and continue checking the other elements.

@HertzDevil Can you come up with a scenario where returning `true` there might give a false positive?

Also, feel free to send a PR for this! Otherwise I can do it.

**Comment by straight-shoota** [comment]
This is the original ticket that introduced this code in Ruby: https://bugs.ruby-lang.org/issues/1448
I'm afraid there's no explanation of the reasoning, just a claim that it works.

But I think it makes sense and actually seems to work. I don't see a reason to doubt it. Though I have no proof either.

**Comment by HertzDevil** [comment]
If we add `#==` then it is natural to also implement a working `#hash` so that recursive `Array`s can be used as `Hash` keys. This works in Ruby:

```ruby
a = []
a << a

b = []
c = [b]
b << c

{a => 1, b => 2}.keys == [c] # => true
```

[Ruby's hasher is also recursive](https://github.com/ruby/ruby/blob/c53bdb8ff64ad707ebbb0c05823c1cb0376433dc/hash.c#L198). Calling `a.hash` in the original example leads to a stack overflow.

I am thinking if we could make `#==` work on _all_ reference types without having to repeat `exec_recursive` in the method body.

**Comment by asterite** [comment]
> I am thinking if we could make #== work on all reference types without having to repeat exec_recursive in the method body

Reference types are compared by reference, so there's no possibility of recursion there.

However, there's the `Object.def_equals` macro that defines `==` by comparing the given fields. I'm thinking we could change that macro to take recursion into account if `self` is a `Reference`. And if we check recursion in `hash` we can do the same in `def_hash`.

Then we have structs where `==` is done by comparing all fields, but I think there's no possibility of recursion there because structs are value types so they can't be repeated.

**Comment by HertzDevil** [comment]
Yeah `def_equals` is what I was thinking of; a shard may anticipate its reference types (with `#==` overridden) to be used in self-referential data structures, or those types may already be self-referential already, yet `exec_recursive` isn't even public. This is the concrete example I came up with a while ago:

```crystal
class Foo
  getter x = [] of Foo
  def_equals_and_hash x
end

foo = Foo.new
foo.x << foo
bar = Foo.new
bar.x << bar
foo == bar # Stack overflow (e.g., infinite or very deep recursion)
```

**Comment by asterite** [comment]
I think for `hash` we can also use the `exec_recursive` we have. In case of recursion we return a default hash value. I think that's what Ruby does. That said, I don't know why they use a different `rb_exec_recursive_outer` function.

**Comment by straight-shoota** [comment]
An implementation has already been proposed in #9000
